File: lib-segment/src/main/java/websiteschema/mpsegment/dict/domain/DomainDictionary.java
package websiteschema.mpsegment.dict.domain;

import org.apache.log4j.Logger;
import websiteschema.mpsegment.dict.*;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.Iterator;
import java.util.List;

public class DomainDictionary implements IDictionary {

    public DomainDictionary() {
        wordNameIndexHashMap = new HashMap<String, Integer>();
        arrayWordItem = new ArrayList<DomainWordItem>();
        synonymIndexHashMap = new HashMap<Integer, Integer>();
        synonymHashMap = new HashMap<Integer, List<Integer>>();
        hashDictionary = new HashDictionary<DomainWordItem>();
        hashDictionary.setHeadLength(2);
    }

    private DomainWordItem lookupWord(String wordName) {
        return hashDictionary.lookupWord(wordName);
    }

    private DomainWordItem getWord(int wordIndex) {
        return wordIndex >= 0 ? this.arrayWordItem.get(wordIndex) : null;
    }

    private int getWordIndex(String wordName) {
        return wordNameIndexHashMap.containsKey(wordName)
                ? wordNameIndexHashMap.get(wordName) : -1;
    }

    private int addWord(String wordName, String pos, int freq, int domainType) throws DictionaryException {
        DomainWordItem word = new DomainWordItem(wordName, domainType);
        word.setOccuredCount(pos, freq);
        int index = arrayWordItem.size();
        wordNameIndexHashMap.put(wordName, index);
        arrayWordItem.add(word);
        hashDictionary.addWord(word);
        if (wordName.length() > maxWordLength) {
            maxWordLength = wordName.length();
        }
        return index;
    }

    public void pushWord(String wordName, String synonym, String pos, int freq, int domainType) throws DictionaryException {
        if (wordName == null || wordName.trim().equals("") || pos == null || pos.trim().equals("") || wordName.trim().length() < 2) {
            l.warn("Load a error word '" + wordName + "'into domain dictionary, already ignored.");
            return;
        }
        wordName = wordName.trim();
        DomainWordItem word = lookupWord(wordName);
        int index = -1;
        if (null != word) {
            word.setOccuredCount(pos, freq);
            word.setDomainType(domainType);
            index = getWordIndex(wordName);
        } else {
            index = addWord(wordName, pos, freq, domainType);
        }
        if (null != synonym && !synonym.isEmpty()) {
            int synonymIndex = getWordIndex(synonym);
            addSynonym(index, synonymIndex);
        }
    }

    private void addSynonym(int index, int synonymIndex) {
        if (index >= 0 && synonymIndex >= 0) {
            this.synonymIndexHashMap.put(index, synonymIndex);
            List<Integer> synonymSet = this.synonymHashMap.get(synonymIndex);
            if (null == synonymSet) {
                synonymSet = new ArrayList<Integer>();
                synonymHashMap.put(synonymIndex, synonymSet);
            }
            if (!synonymSet.contains(index)) {
                synonymSet.add(index);
            }
        }
    }

    public List<IWord> getSynonymSet(String wordName) {
        int index = getWordIndex(wordName);
        if (index >= 0) {
            int synonymIndex = synonymIndexHashMap.containsKey(index) ? synonymIndexHashMap.get(index) : -1;
            List<Integer> synonymSet = null;
            if (synonymIndex >= 0) {
                synonymSet = synonymHashMap.get(synonymIndex);
            } else {
                synonymSet = synonymHashMap.get(index);
            }
            if (null != synonymSet) {
                List<IWord> ret = new ArrayList<IWord>();
                DomainWordItem head = getWord(synonymIndex >= 0 ? synonymIndex : index);
                if (null != head) {
                    ret.add(head);
                }
                for (int i : synonymSet) {
                    ret.add(getWord(i));
                }
                return ret;
            }
        }
        return null;
    }

    @Override
    public IWord getWord(String wordName) {
        return hashDictionary.getWord(wordName);
    }

    public IWord getWordItem(String wordName, int minLength) {
        int l1 = maxWordLength;
        if (wordName.length() < l1) {
            l1 = wordName.length();
        }
        DomainWordItem word = null;
        for (int i2 = l1; i2 >= minLength; i2--) {
            String s2 = wordName.substring(0, i2);
            int wordIndex = getWordIndex(s2);
            if (wordIndex <= 0) {
                continue;
            }
            word = getWord(wordIndex);
            break;
        }

        return word;
    }

    @Override
    public IWord[] getWords(String wordStr) {
        return hashDictionary.getWords(wordStr);
    }

    @Override
    public Iterator<IWord> iterator() {
        return (Iterator) arrayWordItem.iterator();
    }

    private HashDictionary<DomainWordItem> hashDictionary;
    private HashMap<String, Integer> wordNameIndexHashMap;
    private List<DomainWordItem> arrayWordItem;
    private HashMap<Integer, Integer> synonymIndexHashMap;
    private HashMap<Integer, List<Integer>> synonymHashMap;
    private int maxWordLength = 0;
    private final static Logger l = Logger.getLogger("segment");
}

File: lib-segment/src/main/scala/me/yingrui/segment/dict/domain/DomainDictionary.scala
package me.yingrui.segment.dict.domain

import me.yingrui.segment.dict._

import collection.mutable

class DomainDictionary extends IDictionary {
  private val hashDictionary = new HashDictionary()
  hashDictionary.setHeadLength(2)
  private val wordNameIndexHashMap = mutable.HashMap[String, Int]()
  private var arrayWordItem = List[DomainWordItem]()
  private val synonymIndexHashMap = mutable.HashMap[Int, Int]()
  private val synonymHashMap = mutable.HashMap[Int, List[Int]]()
  private var maxWordLength: Int = 0

  override def getWord(wordName: String): IWord = {
    val words = hashDictionary.getWords(wordName)
    if (words != null && !words.isEmpty) words.head else null
  }

  override def getWords(wordStr: String): Array[IWord] = {
    return hashDictionary.getWords(wordStr)
  }

  override def iterator(): List[IWord] = {
    return arrayWordItem
  }

  override def addWord(word: IWord) {
    addWord(word.getWordName(),
      POSUtil.getPOSString(word.getWordPOSTable()(0)(0)),
      word.getWordPOSTable()(0)(1), word.getDomainType())
  }

  override def lookupWord(wordName: String): IWord = {
    return hashDictionary.lookupWord(wordName)
  }

  def pushWord(wordName: String, synonym: String, pos: String, freq: Int, domainType: Int) {
    if (wordName == null || wordName.trim().equals("") || pos == null || pos.trim().equals("") || wordName.trim().length() < 2) {
      System.err.println("Load a error word '" + wordName + "'into domain dictionary, already ignored.")
      return
    }

    val word = lookupWord(wordName)
    var index = -1
    if (null != word) {
      word.setOccuredCount(pos, freq)
      word.setDomainType(domainType)
      index = getWordIndex(wordName)
    } else {
      index = addWord(wordName, pos, freq, domainType)
    }
    if (null != synonym && !synonym.isEmpty()) {
      val synonymIndex = getWordIndex(synonym)
      addSynonym(index, synonymIndex)
    }
  }

  def getSynonymSet(wordName: String): List[IWord] = {
    val index = getWordIndex(wordName)
    if (index >= 0) {
      val synonymIndex = synonymIndexHashMap.getOrElse(index, -1)
      val synonymSet: List[Int] =
        if (synonymIndex >= 0) synonymHashMap(synonymIndex) else synonymHashMap(index)
      if (null != synonymSet) {
        val head = getWord(if (synonymIndex >= 0) synonymIndex else index)
        return head :: synonymSet.map(i => getWord(i))
      }
    }
    return null
  }

  private def getWordIndex(wordName: String): Int = wordNameIndexHashMap.getOrElse(wordName, -1)

  private def getWord(wordIndex: Int): IWord = {
    return if (wordIndex >= 0) arrayWordItem(wordIndex) else null
  }

  private def addWord(wordName: String, pos: String, freq: Int, domainType: Int): Int = {
    val word = DomainWordItem(wordName, domainType)
    word.setOccuredCount(pos, freq)

    addWordPOS(word)

    val index = arrayWordItem.size
    wordNameIndexHashMap += (wordName -> index)
    arrayWordItem = arrayWordItem ++ List(word)
    hashDictionary.addWord(word)
    if (wordName.length() > maxWordLength) {
      maxWordLength = wordName.length
    }
    return index
  }

  private def addWordPOS(word: IWord) {
    val coreDictionary = DictionaryFactory().getCoreDictionary()
    if (coreDictionary != null) {
      val wordInCoreDictionary = coreDictionary.getWord(word.getWordName())
      if (null != wordInCoreDictionary) {
        val posTable = wordInCoreDictionary.getPOSArray().getWordPOSTable()
        for (posAndFreq <- posTable) {
          word.setOccuredCount(POSUtil.getPOSString(posAndFreq(0)), posAndFreq(1))
        }
      }
    }
  }

  private def addSynonym(index: Int, synonymIndex: Int) {
    if (index >= 0 && synonymIndex >= 0) {
      synonymIndexHashMap += (index -> synonymIndex)
      var synonymSet = synonymHashMap.getOrElse(synonymIndex, null)
      if (null == synonymSet) {
        synonymSet = List[Int]()
      }
      if (!synonymSet.contains(index)) {
        synonymSet = synonymSet ++ List(index)
      }
      synonymHashMap += (synonymIndex -> synonymSet)
    }
  }
}
